### PR TITLE
Auto guess encoding

### PR DESCRIPTION
**Pull request by katainaka0503**
This related to #5388.
This PR is modified one of #10013.

From #10013,
Added `'files.autoDetectEncoding'` in `settings.json`  to decide whether to automatically detect encoding.

### CODE REVIEW

**Comment by buzzzzer** [comment]
@katainaka0503 
could be better used files.encoding = 'auto' instead of files.autoDetectEncoding ?

**Comment by katainaka0503** [comment]
@buzzzzer 
I had considered that before I implemented this. And I realized that when opening a new file, VSCode can't specify preferred encoding if `'files.encoding'` is set to `'auto'`.

**Comment by bpasero** [comment]
Thanks for this PR, current status is that we are reviewing the implications of JSChardets LGPL license for VS Code.

**Comment by katainaka0503** [comment]
Rebased.

File: src/vs/base/node/encoding.ts
Comment (vzaritovsky) [inline src/vs/base/node/encoding.ts]: This meant to be "utf-32" instead of "urf-32", right?

**Review by vzarytovskii — CHANGES_REQUESTED** [review]

File: src/vs/base/node/encoding.ts
Comment (katainaka0503) [inline src/vs/base/node/encoding.ts]: Yes. Just a typo. I am so sorry.

Comment (katainaka0503) [inline src/vs/base/node/encoding.ts]: Fixed.

**Comment by LiPengfei19820619** [comment]
I  expect this feature to be merged into VSCode. Thanks very much.

**Comment by bpasero** [comment]
No status update yet from what I wrote [here](https://github.com/Microsoft/vscode/pull/21416#issuecomment-282710795).

**Comment by LiPengfei19820619** [comment]
Any progress? Thanks.

**Comment by katainaka0503** [comment]
I think it is no use to rush.

**Comment by katainaka0503** [comment]
Rebased

**Comment by bpasero** [comment]
@katainaka0503 have you thought about allowing a user to ad-hoc detect the encoding of an opened file from the encoding picker? 

From a file, click on the encoding and pick the option to Reopen with Encoding:
image

In this list, the first entry (maybe with a separator) could be to automatically detect the encoding from the file:
image

One thing to keep in mind is that currently we only allow the first 512 bytes of the file to be used when detecting the encoding. This might cause some encodings to not get detected properly. If we have an explicit action for detection, we could maybe increase the buffer size to give the detection a higher chance of finding an encoding.

Otherwise this PR seems well programmed, good job. Still waiting on the license approval.

**Comment by katainaka0503** [comment]
@bpasero Thanks! I am very pleased.

I think it's very good if it is able to select `automatically detect encoding` from encoding picker.
But I don't like to select the entry every time I open the file.

I think one of the problems is the word `detect` implies vscode can detect the encoding properly every time, so maybe the word `guess` should be used instead for example.

And also In the description of `files.autoDetectEncoding`  ,  it's should be written that `Files may be detected inappropriately.` for example.

What's do you think?

**Comment by bpasero** [comment]
@katainaka0503 yes, this would be an optional thing to "guess" the encoding from the picker in case the user does not want to opt in to the setting for all files. Since we already have jschardet in the system, this functionality could be added quite easily I guess.

As for the setting, maybe it would be better to rename it to `autoGuessEncoding` and have the wording be _When enabled, will attempt to guess the character set encoding when opening files._

I would then also suggest to rename the flag from `autoDetectEncoding` to `autoGuessEncoding`.

As you probably already figured out, we are using [iconv-lite](https://github.com/ashtuchkin/iconv-lite) to convert to and from encodings. If jschardet is returning an encoding with a specific name, we have to ensure that this is the same name used in iconv-lite (see [here](https://github.com/ashtuchkin/iconv-lite/wiki/Supported-Encodings) for a list of supported encodings in iconv-lite). If the names are not matching, the check for [`encodingExists`](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/encoding.ts#L39) would fail. 

Finally I am running into some issues when testing this and I think this exposes a problem with encoding auto detection in general. After making changes to a file and reopening it, the detected encoding is suddenly different and characters cannot be displayed anymore. Steps to reproduce:
* create a file with the contents below saved as `cp1250` encoding 
* open the file in VS Code with encoding detection, it gets opened as `ISO-8859-2`
* add the characters `öäü` to the end of the file in a newline
* reopen the file and notice how the encoding is now detected as `windows-1255`

Contents of file:
```
// A kódolás szinte minden adatátviteli és kommunikációs rendszerben használható, és a következõ európai nyelvek megjelenítésére alkalmas: bosnyák, cseh, horvát, lengyel, magyar, román, szerb (a latinbetÿs írásmóddal), szerbhorvát, szlovák, szlovén, alsó-szorb és felsõ-szorb.
```

After reopening:

image

**Comment by katainaka0503** [comment]
@bpasero  Ok, I understand.
And I'll implement this.

Then, the remaining tasks are

- [ ] Fix description.
- [ ] Rename autoDetectEncoding to autoGuessEncoding.

- [ ] Add autoGuessEncoding entry in encoding picker.

Thanks for precise test!

**Comment by katainaka0503** [comment]
@bpasero I found there are two ways to add entry in encoding picker.

One is to add `autoGuessEncoding` to encoding picker.
The another is to add `Shift_JIS(Guessed from content)` or `ISO-8859-2(Guessed from content)`. 

I think the latter is better. Because users can decide based result of guessing encoding.

**Comment by bpasero** [comment]
@katainaka0503 agreed that showing the guessed encoding is more helpful.

Have you considered my other feedback:
* encoding names have to match what iconv-lite expects
* encoding can be guessed weirdly when content changes

On the last point, maybe we can tweak this by configuring jschardet to only return encodings with a higher confidence level?

**Comment by katainaka0503** [comment]
@bpasero I'll research difference of encoding names and whether it is able to configure jschardet.

**Comment by katainaka0503** [comment]
Now I fixed to be able to set MINIMUM_THRESHOLD.
It's should be determined how much this value should be.

**Comment by katainaka0503** [comment]
@bpasero 
> Finally I am running into some issues when testing this and I think this exposes a problem with encoding auto detection in general. After making changes to a file and reopening it, the detected encoding is suddenly different and characters cannot be displayed anymore. Steps to reproduce:
>
> create a file with the contents below saved as cp1250 encoding (I was using Sublime Text for this)
open the file in VS Code with encoding detection, it gets opened as ISO-8859-2
add the characters öäü to the end of the file in a newline
reopen the file and notice how the encoding is now detected as windows-1255
Contents of file:
>
> // A kódolás szinte minden adatátviteli és kommunikációs rendszerben használható, és a következõ európai nyelvek megjelenítésére alkalmas: bosnyák, cseh, horvát, lengyel, magyar, román, szerb (a latinbetÿs írásmóddal), szerbhorvát, szlovák, szlovén, alsó-szorb és felsõ-szorb.
After reopening:

Environment

- OS X version 10.12.3

Sorry, I couldn't reproduce all of the issues.
When I try to save file with cp1250 with Sublime I have an error 



Using VScode instead of sublime text, I have only the issue that the file opened with ISO-8859-2, and not opened with windows-1255 after add.

**Comment by bpasero** [comment]
@katainaka0503 interesting that explains a few things, so I was adding characters that are not supported by the encoding and so it changed. Makes sense I guess.

Did you have any success in changing the threshold? Is there any documentation what one should set it to?

**Comment by katainaka0503** [comment]
@bpasero I tried with threshold 0.95, but after windows1250 stored (some characters replaced with `?` in some place), then opened file, it remained to be detected as ISO-8859-2.

I think we should broke this in two cases.

- What encoding should be selected when the content is interpreted as same content in more than one encodings. This occured in above case (after characters not supported was replaced.) Windows1250 and ISO-8859-2
    Probably, this is hard to be solved with high threshold level.

- The same with first but interpreted differently in those encodings
    I think this case can be solved with high threshold level. Maybe this is already solved with current threshold.

And I could not found any information about threshold value.

**Comment by bpasero** [comment]
Maybe we go with what we have now and wait for more user information on how this behaves in the specific cases.

**Comment by katainaka0503** [comment]
I added a guessed encoding entry in encoding picker as a prototype

**Comment by bpasero** [comment]
Merged. did some slight changes to encoding picker to prevent showing encodings multiple times.

**Comment by katainaka0503** [comment]
@bpasero Thanks!

**Comment by bpasero** [comment]
@katainaka0503 just so that you know, there is a number of cases already where encoding guessing does not work: https://github.com/aadsm/jschardet/issues/31, https://github.com/aadsm/jschardet/issues/29, https://github.com/aadsm/jschardet/issues/30

**Comment by katainaka0503** [comment]
@bpasero Thanks for letting me know! I'll check them.

**Pull request merged**
